// bits-ui/dist/bits/checkbox/checkbox.svelte.js
import { srOnlyStyles, styleToString, useRefById } from "svelte-toolbelt";
import { watch } from "runed";
import { getAriaChecked, getAriaRequired, getDataDisabled } from "../../internal/attrs.js";
import { kbd } from "../../internal/kbd.js";
import { createContext } from "../../internal/create-context.js";
const CHECKBOX_ROOT_ATTR = "data-checkbox-root";
const CHECKBOX_GROUP_ATTR = "data-checkbox-group";
const CHECKBOX_GROUP_LABEL_ATTR = "data-checkbox-group-label";
class CheckboxGroupState {
    id;
    ref;
    value;
    disabled;
    required;
    name;
    labelId = $state(undefined);
    constructor(props) {
        this.id = props.id;
        this.ref = props.ref;
        this.value = props.value;
        this.disabled = props.disabled;
        this.required = props.required;
        this.name = props.name;
        useRefById({
            id: this.id,
            ref: this.ref,
        });
    }
    addValue(checkboxValue) {
        if (!checkboxValue)
            return;
        if (!this.value.current.includes(checkboxValue)) {
            this.value.current.push(checkboxValue);
        }
    }
    removeValue(checkboxValue) {
        if (!checkboxValue)
            return;
        const index = this.value.current.indexOf(checkboxValue);
        if (index === -1)
            return;
        this.value.current.splice(index, 1);
    }
    props = $derived.by(() => ({
        id: this.id.current,
        role: "group",
        "aria-labelledby": this.labelId,
        "data-disabled": getDataDisabled(this.disabled.current),
        [CHECKBOX_GROUP_ATTR]: "",
    }));
}
class CheckboxGroupLabelState {
    id;
    ref;
    group;
    constructor(props, group) {
        this.id = props.id;
        this.ref = props.ref;
        this.group = group;
        useRefById({
            id: this.id,
            ref: this.ref,
            onRefChange: (node) => {
                if (node) {
                    group.labelId = node.id;
                }
                else {
                    group.labelId = undefined;
                }
            },
        });
    }
    props = $derived.by(() => ({
        id: this.id.current,
        "data-disabled": getDataDisabled(this.group.disabled.current),
        [CHECKBOX_GROUP_LABEL_ATTR]: "",
    }));
}
class CheckboxRootState {
    #id;
    #ref;
    #type;
    checked;
    #disabled;
    #required;
    #name;
    value;
    indeterminate;
    group = null;
    trueName = $derived.by(() => {
        if (this.group && this.group.name.current) {
            return this.group.name.current;
        }
        else {
            return this.#name.current;
        }
    });
    trueRequired = $derived.by(() => {
        if (this.group && this.group.required.current) {
            return true;
        }
        return this.#required.current;
    });
    trueDisabled = $derived.by(() => {
        if (this.group && this.group.disabled.current) {
            return true;
        }
        return this.#disabled.current;
    });
    constructor(props, group = null) {
        this.checked = props.checked;
        this.#disabled = props.disabled;
        this.#required = props.required;
        this.#name = props.name;
        this.value = props.value;
        this.#ref = props.ref;
        this.#id = props.id;
        this.indeterminate = props.indeterminate;
        this.#type = props.type;
        this.group = group;
        this.onkeydown = this.onkeydown.bind(this);
        this.onclick = this.onclick.bind(this);
        useRefById({
            id: this.#id,
            ref: this.#ref,
        });
        watch([() => $state.snapshot(this.group?.value.current), () => this.value.current], ([groupValue, value]) => {
            if (!groupValue || !value)
                return;
            this.checked.current = groupValue.includes(value);
        });
        watch(() => this.checked.current, (checked) => {
            if (!this.group)
                return;
            if (checked) {
                this.group?.addValue(this.value.current);
            }
            else {
                this.group?.removeValue(this.value.current);
            }
        });
    }
    onkeydown(e) {
        if (this.#disabled.current)
            return;
        if (e.key === kbd.ENTER)
            e.preventDefault();
        if (e.key === kbd.SPACE) {
            e.preventDefault();
            this.#toggle();
        }
    }
    #toggle() {
        if (this.indeterminate.current) {
            this.indeterminate.current = false;
            this.checked.current = true;
        }
        else {
            this.checked.current = !this.checked.current;
        }
    }
    onclick(_) {
        if (this.#disabled.current)
            return;
        this.#toggle();
    }
    snippetProps = $derived.by(() => ({
        checked: this.checked.current,
        indeterminate: this.indeterminate.current,
    }));
    props = $derived.by(() => ({
        id: this.#id.current,
        role: "checkbox",
        type: this.#type.current,
        disabled: this.trueDisabled,
        "aria-checked": getAriaChecked(this.checked.current, this.indeterminate.current),
        "aria-required": getAriaRequired(this.trueRequired),
        "data-disabled": getDataDisabled(this.trueDisabled),
        "data-state": getCheckboxDataState(this.checked.current, this.indeterminate.current),
        [CHECKBOX_ROOT_ATTR]: "",
        //
        onclick: this.onclick,
        onkeydown: this.onkeydown,
    }));
}
//
// INPUT
//
class CheckboxInputState {
    root;
    trueChecked = $derived.by(() => {
        if (this.root.group) {
            if (this.root.value.current !== undefined &&
                this.root.group.value.current.includes(this.root.value.current)) {
                return true;
            }
            return false;
        }
        return this.root.checked.current;
    });
    shouldRender = $derived.by(() => Boolean(this.root.trueName));
    constructor(root) {
        this.root = root;
    }
    props = $derived.by(() => ({
        type: "checkbox",
        checked: this.root.checked.current === true,
        disabled: this.root.trueDisabled,
        required: this.root.trueRequired,
        name: this.root.trueName,
        value: this.root.value.current,
        "aria-hidden": "true",
        style: styleToString(srOnlyStyles),
    }));
}
//
// HELPERS
//
function getCheckboxDataState(checked, indeterminate) {
    if (indeterminate) {
        return "indeterminate";
    }
    return checked ? "checked" : "unchecked";
}
//
// CONTEXT METHODS
//
const [setCheckboxGroupContext, getCheckboxGroupContext] = createContext("Checkbox.Group");
const [setCheckboxRootContext, getCheckboxRootContext] = createContext("Checkbox.Root");
export function useCheckboxGroup(props) {
    return setCheckboxGroupContext(new CheckboxGroupState(props));
}
export function useCheckboxRoot(props) {
    return setCheckboxRootContext(new CheckboxRootState(props, getCheckboxGroupContext(null)));
}
export function useCheckboxGroupLabel(props) {
    return new CheckboxGroupLabelState(props, getCheckboxGroupContext());
}
export function useCheckboxInput() {
    const root = getCheckboxRootContext();
    return new CheckboxInputState(root);
}
